Add test runner library

aakash070 opened the pull request:
This change adds a test runner library under tools/celtest which provides native language support for setting up and executing tests. 
1. The library makes use of the compiler tool for setting up the CEL runtime environment.
2. It also provides callback options for augmenting the test runner with test suite path, test suite parser, file descriptor set path.

The change also includes a test harness in YAML which aligns with cel.expr.conformance.test.TestSuite.
The following changes will be raised separately:
1. The bazel rule to trigger the test runner with parameterized options
2. Sample usages for triggering test runner library for policies, expressions and compiled expressions.

TristonianJones commented on test/suite.go:
```diff
@@ -0,0 +1,60 @@
+type Case struct {
+	Name        string `yaml:"name"`
+	Description string `yaml:"description"`
+	Input       Input  `yaml:"input"`
```
Let's shift this to be `Input map[string]InputValue` and add a second field called `InputContext` which refers to a struct containing the `ContextExpr string`:

```go
type Case struct {
        Name        string `yaml:"name"`
	Description string `yaml:"description"`
	Input       map[string]*InputValue  `yaml:"input"`
	InputContext *InputContext `yaml:"input_context"`
        Output *Output `yaml:"output"`
}
```

```go
type InputContext struct {
   ContextExpr string `yaml:"context_expr"
}
```

aakash070 replied:
Do we need a separate struct to contain the `ContextExpr` string or is it to maintain parity with the textproto schema. We could also move this up along side the input map:
```
type Case struct {
        Name        string `yaml:"name"`
	Description string `yaml:"description"`
	Input       map[string]*InputValue  `yaml:"input"`
	InputContext string `yaml:"input_context"`
        Output *Output `yaml:"output"`
}
```

TristonianJones replied:
@aakash070 I kept the same fields for parity between the two different file formats. Context inputs are the less common case. Another alternative is to keep the `InputContext` and mark it as inline which would hoist the `context_expr` field up without making users type `input_context` first:

```go
type Case struct {
        Name        string `yaml:"name"`
	Description string `yaml:"description"`
	Input       map[string]*InputValue  `yaml:"input,omitempty"`
	InputContext *InputContext `yaml:",inline,omitemtpy"`
        Output *Output `yaml:"output"`
}

type InputContext struct {
   ContextExpr string `yaml:"context_expr"`
}
```

```yml
# Populates a `Case` object with an `InputContext` struct where
# `ContextExpr` is set to the expression
test_case:
  name: ...
  context_expr: "qualified.ContextMsg{...}"
  output: ...
```

TristonianJones commented on tools/celtest/test_runner.go:
```diff
@@ -0,0 +1,698 @@
+// TestResult represents the result of a test case execution. It contains the validation result
+// along with the expected result and any errors encountered during the execution.
+type TestResult struct {
+	Success bool
```
Does success mean the execution completed successfully or that the result matcher condition was satisfied? Consider adding a few comments on the fields to explain how they behave

TristonianJones reviewed (changes requested):
Please add test cases for the `test_runner.go` and `compiler.go` as part of this PR. Thank you!

aakash070 commented on tools/celtest/test_runner.go:
```diff
@@ -0,0 +1,698 @@
+
+// DefaultTestSuiteParserTextproto is a default parser for a test suite file serialized in a
+// Textproto file. The test suite is a cel.expr.conformance.test.TestSuite message.
+func DefaultTestSuiteParserTextproto(t *testing.T, path any) (*conformancepb.TestSuite, error) {
```
Thanks for pointing, yes, these can be made private methods now. 
I was thinking about a specific case when I made these as public methods, where the tests would be triggered by providing Compiler and Test Runner options programmatically rather than via the build rule. In such a case, the test suite parser would also need to be configured via a Test Runner option. But, I updated `TestSuiteParser` to handle such cases where the test suite file contains messages/objects in standard format and missed changing these back to private methods.

aakash070 commented on tools/celtest/test_runner.go:
```diff
@@ -0,0 +1,698 @@
+	Expressions              []compiler.InputExpression
+	TestSuiteFilePath        string
+	FileDescriptorSetPath    string
+	TestSuiteParserTextproto TestSuiteParserTextproto
```
We may need to allow the parsers to remain configurable by the users. This can help handle cases where the test suite is:
1. Either derived from a Textproto/YAML file whose contents are not in the standard test suite format, i.e. neither a cel.spec.expr.conformance.test.TestSuite message nor a test.TestSuite object.
2. Or created dynamically by the users at runtime. For example: there are cases where the users want to generate the test suite by leveraging the data stored in their db rather than maintaining test suite files.

aakash070 commented on tools/celtest/test_runner.go:
```diff
@@ -0,0 +1,698 @@
+
+// TestSuiteParserTextproto is a functional interface that provides a cel.expr.conformance.test.TestSuite
+// message. The message can be serialized in a Textproto file or derived programmatically.
+type TestSuiteParserTextproto func(*testing.T, any) (*conformancepb.TestSuite, error)
```
This refers to cases where `conformancepb.TestSuite` is generated dynamically rather than being stored in file format. This would permit the users to perform operations like db lookups or invoke rpcs if needed to create the test suite.

aakash070 commented on tools/celtest/test_runner.go:
```diff
@@ -0,0 +1,698 @@
+// - set the test suite parser based on the file format: YAML or Textproto
+type TestRunnerOption func(*TestRunner) (*TestRunner, error)
+
+// TestSuiteParserTextproto is a functional interface that provides a cel.expr.conformance.test.TestSuite
```
Updated the documentation

TristonianJones commented on tools/celtest/test_runner.go:
```diff
@@ -0,0 +1,708 @@
+
+// ParseTextproto parses a test suite file in Textproto format.
+func (p *tsParser) ParseTextproto(path any) (*conformancepb.TestSuite, error) {
+	filePath := path.(string)
```
If the path only works as a string, should it be switched from `any` to `string` in the interface and implementations?

aakash070 replied:
I had kept this as `any` to allow for custom implementations of the interface to pass non-string input args to the methods. But that requirement can be handled later when needed. Will make this change.

TristonianJones commented on tools/celtest/test_runner.go:
```diff
@@ -0,0 +1,708 @@
+	}
+	var programs []cel.Program
+	for _, expr := range tr.Expressions {
+		ast, _, err := expr.CreateAST(tr.Compiler)
```
Can we add a TODO to propagate the metadata from CreateAST back with the Program instances? Possibly as another struct which combines this information?

TristonianJones commented on tools/celtest/test_runner.go:
```diff
@@ -0,0 +1,708 @@
+	if err != nil {
+		return nil, fmt.Errorf("e.Extend() failed: %v", err)
+	}
+	ast, iss := e.Compile(expr)
```
You can probably wrap the `CreateEnv` and extension with optional types into a `sync.DoOnce` to avoid building it each time

aakash070 commented:
@TristonianJones I have added tests for `test_runner.go`. Tests existed for `compiler.go`.
I have moved one of the tests for custom policy metadata options from compiler to test_runner. Please check.

TristonianJones commented on tools/celtest/test_runner.go:
```diff
@@ -0,0 +1,707 @@
+	}
+}
+
+func testRunnerOptions(testRunnerOpts []TestRunnerOption, testCompilerOpts ...any) []TestRunnerOption {
```
Could you actually make this a functional option to configure common options from flags? Can be done in a follow-up

```go
func TestOptionsFromFlags() TestRunnerOption {
   return func(tr *TestRunner) (*TestRunner, error) {
      opts := []TestRunnerOption{
          DefaultTestSuiteParser(testSuitePath),
          AddFileDescriptorSet(...),
          ...
       }
       var err error
       for _, o := range opts {
         tr, err = o(tr)
         if err != nil {
           return nil, err
         }
       }
       return tr
   }
}

TristonianJones replied:
Resolving for now, in order to merge, but please update in the next PR
